src/lib.rs

```diff
@@ -448,6 +448,30 @@ pub fn optimize_from_memory(data: &[u8], opts: &Options) -> PngResult<Vec<u8>> {
     }
 }
 
+/// Perform optimization on the raw image data using the options provided
+pub fn optimize_from_raw(raw: PngImage, opts: &Options) -> PngResult<Vec<u8>> {
+    info!("Processing from raw image data");
+
+    let bpp = (raw.ihdr.bit_depth.as_u8() * raw.channels_per_pixel()) as usize;
+    let row_bytes = (bpp * raw.ihdr.width as usize + 7) / 8;
+    let expected_len = row_bytes * raw.ihdr.height as usize;
+    if raw.data.len() != expected_len {
+        return Err(PngError::new(&format!(
+            "Data length {} does not match the expected length {}",
+            raw.data.len(),
+            expected_len
+        )));
+    }
+
+    let deadline = Arc::new(Deadline::new(opts.timeout));
+
+    if let Some(png) = optimize_raw(Arc::new(raw), opts, deadline, None) {
```

Comment:
Regarding your question about making `optimize_from_raw` take `PngImage` by reference, I don't think we should do that. Like `Box`es, `Arc`s own the data they point to, and you can't own a reference, so you would have to clone the image. Therefore, even if such a refactor can be done without throwing ergonomics out the window, I think it's out of scope for this PR.

Instead, I propose refactoring `optimize_from_raw` to generically take anything that can be converted to an `Arc<PngImage>`:

```suggestion
pub fn optimize_from_raw<T: Into<Arc<PngImage>>>(raw: T, opts: &Options) -> PngResult<Vec<u8>> {
    info!("Processing from raw image data");

    let raw = raw.into();

    let bpp = (raw.ihdr.bit_depth.as_u8() * raw.channels_per_pixel()) as usize;
    let row_bytes = (bpp * raw.ihdr.width as usize + 7) / 8;
    let expected_len = row_bytes * raw.ihdr.height as usize;
    if raw.data.len() != expected_len {
        return Err(PngError::new(&format!(
            "Data length {} does not match the expected length {}",
            raw.data.len(),
            expected_len
        )));
    }

    let deadline = Arc::new(Deadline::new(opts.timeout));

    if let Some(png) = optimize_raw(raw, opts, deadline, None) {
```

This refactor is relatively simple in comparison, and allows client code to avoid cloning the `PngImage` if it already has it inside an `Arc`, which is likely to be a common case.

Reply:
Great idea, thanks!